File: dissertation-frontend/lambdas/getWorkstationsByPlantIndex.js
const AWS = require("aws-sdk");

// Set up the DynamoDB client
const dynamodb = new AWS.DynamoDB.DocumentClient();

exports.handler = async (event) => {
  try {
    // Get the plantIndex from the URL path parameter
    const plantIndex = event.pathParameters.plantIndex;

    // Define the parameters for the DynamoDB scan operation
    const params = {
      TableName: "workstations",
    };

    // Use the DynamoDB scan operation to retrieve all items from the table
    const result = await dynamodb.scan(params).promise();

    // Filter the results based on the plantIndex attribute
    const workstations = result.Items.filter(
      (item) => item.plantIndex === plantIndex
    );

    // Return the filtered list of workstations as the response
    return {
      statusCode: 200,
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Headers": "*",
        "Access-Control-Allow-Methods": "OPTIONS,GET",
      },
      body: workstations,
    };
  } catch (error) {
    return {
      statusCode: 500,
      body: error.message,
    };
  }
};

File: dissertation-frontend/lambdas/getWorkstationsByPlantIndex.py
import json
from decimal import Decimal
import boto3

dynamodb = boto3.resource('dynamodb')
table_name = 'workstations'
table = dynamodb.Table(table_name)


def handler(event, context):
    try:
        plant_index = event['plantIndex']
    except:
        return {
            'statusCode': 400,
            'body': json.dumps({'message': 'Missing plantIndex parameter'})
        }

    try:
        response = table.scan(FilterExpression=boto3.dynamodb.conditions.Attr(
            'plantIndex').eq(plant_index))
        items = response['Items']

        # Convert Decimal objects to float or int
        for item in items:
            for key, value in item.items():
                if isinstance(value, Decimal):
                    item[key] = float(value)

    except:
        return {
            'statusCode': 500,
            'body': json.dumps({'message': 'Failed to query DynamoDB'})
        }

    return {
        'statusCode': 200,
        'headers': {
            'Content-Type': 'application/json',
            'Access-Control-Allow-Headers': '*',
            'Access-Control-Allow-Methods': 'OPTIONS,GET'
        },
        'body': items
    }
